contracts/farm/Farm.sol

```diff
@@ -0,0 +1,197 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.17;
+
+import "@openzeppelin/contracts-upgradeable/security/PausableUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+
+import "./libraries/SafeToken.sol";
+import "./IFarm.sol";
+import "../core/interfaces/IERC20.sol";
+
+contract Farm is IFarm, OwnableUpgradeable, PausableUpgradeable {
+    using SafeToken for address;
+
+    struct StakeInfo {
+        uint256 amount;
+        uint256 rewardDebt;
+    }
+
+    struct PoolInfo {
+        uint256 allocPoint;
+        uint256 lastRewardBlock;
+        uint256 accRewardPerShare;
+    }
+
+    address public rewardToken;
+
+    uint256 public rewardPerBlock;
+    uint256 public totalAllocPoint;
```

Comment:
Lets make next improvement - have those details not at contract level, but for each particular pool